### PR TITLE
Pulling nested elements implemented

### LINKED ISSUE #720: Revit_Toolkit: nested elements are not handled correctly
#### Description:

When pulling Revit elements with nested ones, only top level gets pulled - possibly nested elements should be pulled too.

#### Test file(s):

On [SharePoint](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2FRevit%5FToolkit%2DIssue720%2DNestedElements&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4)

### PR DESCRIPTION
**Pull request by enarhi**
### Issues addressed by this PR

Closes #720 

Nested elements are now pulled when pulling from Revit if the pull config option for IncludeNestedElements=True. Default behavior is currently to include nested elements; this can be changed in the future if default nested element pulls prove too heavy, although initial benchmarking indicates minimal extra overhead for most families with nested elements.

### Test files
[Revit and GH Test Files](https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/Revit_Toolkit/Revit_Toolkit-%23720-NestedElements?csf=1&web=1&e=kF7kkq)

### Additional comments
This PR includes implementation for nested instances in windows, which require unique methods due to how window geometry is calculated in revit. Most element types do not require this workaround, but some element types may going forward. These should be covered under separate issues on an as-needed basis.

See images below for examples of implementation, including storage of parent element id to a new revit identifier property. 

_Family with nested elems in Revit_
image

_Pulled in GH showing elem geo for nested elems_
image

_Props in Grasshopper per pull, including parent elem id. THis also shows the pull config with nested elements enabled._
image

### CODE REVIEW

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check code compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check copyright compliance

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
Please fix the documentation compliance issues to begin with :smile:

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project compliance

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
Also CSProject compliance please :smile:

**Comment by enarhi** [comment]
@BHoMBot fix project file ref. 1409293750

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check core

**Comment by enarhi** [comment]
@BHoMBot check compliance

**Comment by enarhi** [comment]
@BHoMBot check core

**Comment by FraserGreenroyd** [comment]
@BHoMBot check core

File: Revit_Core_Engine/Query/OpeningSurface.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Query/OpeningSurface.cs:85]: I do not understand this part of the code - why would you convert the Revit parameters to BHoM parameter objects and then do matching with hard coded names (_Sill Height_, _Head Height_)? This does not look sustainable and I would argue whether it is necessary.

File: Revit_Core_Engine/Query/OpeningSurface.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Query/OpeningSurface.cs:88]: Why do you loop over all hosts if you place `dummyInstance` on the first? By the looks of code, you could use only it further down the line.

File: Revit_Core_Engine/Query/OpeningSurface.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Query/OpeningSurface.cs:110]: This piece of code looks very similar to the existing implementation - couldn't it be wrapped into a private method and called twice to avoid boilerplate?

File: Revit_oM/Parameters/RevitIdentifiers.cs
Comment (pawelbaran) [inline Revit_oM/Parameters/RevitIdentifiers.cs:59]: `ParentElementId` sounds a bit ambiguous to me - what is a parent? I guess people can understand it in various ways: a host, owner view, parent family etc. I would opt for something more precise, like, hm `SuperComponentId`? Or would that be too abstract?

File: Revit_oM/Parameters/RevitIdentifiers.cs
Comment (pawelbaran) [inline Revit_oM/Parameters/RevitIdentifiers.cs:66]: One thing to notice, `RevitIdentifiers` implements `IImmutable` interface, which means it needs to be versioned individually in Versioning_Toolkit - see https://github.com/BHoM/Versioning_Toolkit/pull/104 as an example.

File: Revit_Engine/Create/Config/RevitPullConfig.cs
Comment (pawelbaran) [inline Revit_Engine/Create/Config/RevitPullConfig.cs:56]: Please deprecate the previous version of the method (possibly `PreviousVersionAttribute` could be used, too).

**Review by pawelbaran — CHANGES_REQUESTED** [review]
I did a first round of code review - please see the code comments.

Functionality wise, I have tested it on pile caps and seems to run nicely 👍 However, it would be great if you could provide a few test cases @enarhi so that we could gather them together and build representative test content (windows, doors, pile caps, any other major type that is relevant to that issue).

File: Revit_oM/Parameters/RevitIdentifiers.cs
Comment (enarhi) [reply Revit_oM/Parameters/RevitIdentifiers.cs:66]: noted, will add a pr to versioning accordingly once this is merged up

Comment (pawelbaran) [reply Revit_oM/Parameters/RevitIdentifiers.cs:66]: Coordination will be welcome, otherwise all internalized objects pulled from Revit will change to `BHoMObjects` (that is how I learnt about it 😃).

File: Revit_oM/Parameters/RevitIdentifiers.cs
Comment (enarhi) [reply Revit_oM/Parameters/RevitIdentifiers.cs:59]: Parent / Child adheres to standard nesting naming conventions. I think that is the most appropriate name for it, we could go with HostElement as well but that creates ambiguity between a Host is the nested sense and host in the revit host element sense (eg a wall hosting a window).

File: Revit_Core_Adapter/CRUD/Read.cs
Comment (enarhi) [reply Revit_Core_Adapter/CRUD/Read.cs]: changed

File: Revit_Core_Engine/Query/OpeningSurface.cs
Comment (enarhi) [reply Revit_Core_Engine/Query/OpeningSurface.cs:88]: The usage of the first host instance here is just for placing the dummy instance. The later usage of all hosts is to extract the total opening area resulting from the dummy instances cutout on all hosts in case it overlaps multiple hosts.

File: Revit_Core_Engine/Query/OpeningSurface.cs
Comment (enarhi) [reply Revit_Core_Engine/Query/OpeningSurface.cs:85]: I needed a way to effectively copy parameters, and as far as I could tell this is the best available method so that we can take advantage of the SetParameters method to manage setting of different parameter types. The hardcoded exclusion of Sill Height and Head Height is due to these causing errors if set through the api; they do not need setting anyhow as setting location and the other params places the instance in the exact location without touching these two parameters. If we have a better means of effectively copying revit params from one element and assigning to another I'm all ears!

File: Revit_Core_Engine/Query/OpeningSurface.cs
Comment (enarhi) [reply Revit_Core_Engine/Query/OpeningSurface.cs]: done

**Comment by enarhi** [comment]
@pawelbaran changes made, should be ready for re-review. I'm adding a revit test file as you noted to the pr as well, this will contain only pile caps with piles and nested windows families for now but more to come. One of the nested windows in the test file will not pull nested family geometry through as it currently stands. This is intentionally including to indicate current limitations, as the family that will not pull geometry is not a hosted window and therefore does not have a wall it belongs to or cuts. This would be similar to the equivalent limitation on a non-nested window family that is non-hosted, and should be covered in the future by a separate issue not specific to nested families.

File: Revit_Core_Engine/Query/OpeningSurface.cs
Comment (pawelbaran) [reply Revit_Core_Engine/Query/OpeningSurface.cs:86]: I believe you could simply use in-built Revit get/set parameter functionality - no need to convert to BHoM parameters, it is just an overhead as far as I understand. If you need to skip any parameters, I recommend using [BuiltInParameter enum](https://www.revitapidocs.com/2020/fb011c91-be7e-f737-28c7-3f1e1917a0e0.htm) instead of string.

File: Revit_oM/Parameters/RevitIdentifiers.cs
Comment (pawelbaran) [reply Revit_oM/Parameters/RevitIdentifiers.cs:59]: Hmm, I agree that the name fits into the convention, I am just a bit afraid of ambiguity. But I will not argue any more as I do not have any reasonable alternatives atm - the only thing I would recommend is to highlight that this property will only be filled for nested families (_owns_ suggests the concept of ownership, which may be understood in different ways).

**Comment by pawelbaran** [comment]
I looked at the test file - the functionality makes sense, so my only concerns are with the code itself at the moment. Looking forward to getting the comments resolved and more testing being done by the users (you being a prime one @enarhi 😉).

A forward-reaching thought: it would be nice to have a method that would somehow group the BHoMObjects based on their nesting chain, so that the user could easily find all children and grandchildren of a given parent or simply group all objects that have a common root. How does that sound?

File: Revit_oM/Parameters/RevitIdentifiers.cs
Comment (enarhi) [reply Revit_oM/Parameters/RevitIdentifiers.cs:59]: Agreed, description updated.

File: Revit_Core_Engine/Query/OpeningSurface.cs
Comment (enarhi) [reply Revit_Core_Engine/Query/OpeningSurface.cs]: I've removed the redundant to and from bhom parameter portion. This required boilerplating the switches for the different Revit parameter value types in order to get the param values to then set them. Perhaps worth creating a method that handles that for us? Revit API of course does not provide for something as simple as SetParameter(element, parameter) which would streamline the code here.

Comment (enarhi) [reply Revit_Core_Engine/Query/OpeningSurface.cs]: BuiltInEnum is now used for the Head and Sill Height exclusion as well, good call there!

**Review by pawelbaran — CHANGES_REQUESTED** [review]
A few final comments left, I will also double-test the functionality with the routines that I have used before. Besides that, looks like we are getting there! 👍

**Review by pawelbaran — APPROVED** [review]
Tested, works as expected - LGTM now, both code and functionality wise 👍 Happy to merge as soon as we get a second review from BEnv/MEP (@FraserGreenroyd @michaelhoehn) + Versioning_Toolkit PR is ready as mentioned in [one of my previous comments](https://github.com/BHoM/Revit_Toolkit/pull/906/files#r525076244).

**Comment by enarhi** [comment]
Versioning is ready as referenced above

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Review by FraserGreenroyd — APPROVED** [review]
LGTM

**Pull request merged**
